File: creature.cpp
// creature.cpp
// Implementations for Creature class
// Author: Juan Arias
//
// The Creature class represents something that can move through the Maze
// class, marking locations as visited or part of the path to the exit, and
// returns a string indicating its path in North West South East.

#include "creature.h"

// Friend function
// Overloaded output operator
// Prints the position of the Creature in the format: C(x, y)
std::ostream& operator<<(std::ostream& out, const Creature& creat) {

	out << "C(" << creat.row << ", " << creat.col << ')';

	return out;
}

// Constructs Creature at position in Maze,
// assumes position is valid open space
// Initialies mazePtr to nullptr, row to parameter startRow &
// col to parameter startCol
Creature::Creature(int startRow, int startCol)

	:mazePtr(nullptr), row(startRow), col(startCol) {}

// Destroys Creature
Creature::~Creature() {}

// Returns a string in the form of NNEEN
// (where N means North, E means East, etc)
// that will let the creature get to the exit
// if creature cannot get to the exit, returns "X"
// if constructed at exit, returns ""
// Assumes maze has not been marked
std::string Creature::Solve(Maze& maze) {

	if (maze.IsExit(row, col)) {
		
		maze.MarkAsPath(row, col);
		return "";

	}

	mazePtr = &maze;

	std::string path(goNorth());
	
	path = (path[0] == NO_PATH) ? goWest()  : path;
	path = (path[0] == NO_PATH) ? goEast()  : path;
	path = (path[0] == NO_PATH) ? goSouth() : path;

	if (path[0] == NO_PATH) {

		maze.MarkAsVisited(row, col);
	}

	return path;
}

// Attempts to go North in Maze, updates position &
// returns NORTH if successful, otherwise returns NO_PATH
std::string Creature::goNorth() {

	if (mazePtr->IsClear(row - 1, col)) {
		
		mazePtr->MarkAsPath(row--, col);

		if (mazePtr->IsExit(row, col)) {
		
			mazePtr->MarkAsPath(row, col);

			return std::string(1, NORTH);
		}
		
		std::string newPath(goNorth());

		newPath = (newPath[0] == NO_PATH) ? goWest() : newPath;
		newPath = (newPath[0] == NO_PATH) ? goEast() : newPath;

		if (newPath[0] == NO_PATH) {

			mazePtr->MarkAsVisited(row++, col);
			return newPath;
		}

		return newPath.insert(0, 1, NORTH);
	}

	return std::string(1, NO_PATH);
}

// Attempts to go South in Maze, updates position &
// returns SOUTH if successful, otherwise NO_PATH
std::string Creature::goSouth() {

	if (mazePtr->IsClear(row + 1, col)) {
	
		mazePtr->MarkAsPath(row++, col);

		if (mazePtr->IsExit(row, col)) {
		
			mazePtr->MarkAsPath(row, col);

			return std::string(1, SOUTH);
		
		}

		std::string newPath(goSouth());

		newPath = (newPath[0] == NO_PATH) ? goWest() : newPath;
		newPath = (newPath[0] == NO_PATH) ? goEast() : newPath;

		if (newPath[0] == NO_PATH) {

			mazePtr->MarkAsVisited(row--, col);
			return newPath;

		}

		return newPath.insert(0, 1, SOUTH);
	
	}
	return std::string(1, NO_PATH);

}

// Attempts to go East in Maze, updates position &
// returns EAST if successful, otherwise returns NO_PATH
std::string Creature::goEast() {

	if (mazePtr->IsClear(row, col + 1)) {
	
		mazePtr->MarkAsPath(row, col++);

		if (mazePtr->IsExit(row, col)) {
		
			mazePtr->MarkAsPath(row, col);

			return std::string(1, EAST);
		
		}

		std::string newPath(goEast());

		newPath = (newPath[0] == NO_PATH) ? goNorth() : newPath;
		newPath = (newPath[0] == NO_PATH) ? goSouth() : newPath;

		if (newPath[0] == NO_PATH) {

			mazePtr->MarkAsVisited(row, col--);
			return newPath;

		}

		return newPath.insert(0, 1, EAST);
	}

	return std::string(1, NO_PATH);

}

// Attempts to go West in Maze, updates position &
// returns WEST if successful, otherwise returns NO_PATH
std::string Creature::goWest() {

	if (mazePtr->IsClear(row, col - 1)) {
	
		mazePtr->MarkAsPath(row, col--);

		if (mazePtr->IsExit(row, col)) {
		
			mazePtr->MarkAsPath(row, col);

			return std::string (1, WEST);

		}

		std::string newPath(goWest());

		newPath = (newPath[0] == NO_PATH) ? goNorth() : newPath;
		newPath = (newPath[0] == NO_PATH) ? goSouth() : newPath;

		if (newPath[0] == NO_PATH) {

			mazePtr->MarkAsVisited(row, col++);
			return newPath;

		}

		return newPath.insert(0, 1, WEST);
	
	}

	return std::string(1, NO_PATH);

}

File: Creature.java
/**
 * Creature.java
 * 
 * @author Wizzollz
 * 
 * Solves maze
 */
package css342;

/**
 * Creature mfin class
 * 
 * @author Juan
 * 
 * These objects use exhaustive recursion to solve Maze class objects
 * 		
 * 	  ***************************************************************
 *    *ASSUMES that (0,0) is a valid starting location in given Maze*
 *    ***************************************************************
 */
public class Creature implements MazeSolver {
	
	// Reference to Maze to solve
	private Maze maze;
	
	// Buffer to build solution path String
	private StringBuilder solPath = new StringBuilder();
	
	// Current location of Creature
	private int row = 0, col = 0, solPathIndex = 0;
	
	/**
	 * Solves the given Maze and returns the directional solution path
	 * @param maze The Maze to solve
	 * @return the directional solution path as a String
	 */
	public String solve(Maze maze) {
		
		if (maze.isExit(this.row, this.col)) {
			
			maze.markAsPath(this.row, this.col);
			
			return DIRECTION.NONE.toString();
		}
		
		this.maze = maze;
		
		if (!this.north()) {
			
			if (!this.east()) {
				
				if (!this.west()) {
					
					this.south();
				}
			}
		}
		
		return this.formatPath();
	}
	
	/**
	 * String representation of this hoe
	 * @return String mfin format
	 */
	public String toString() {
		
		return String.format("C(%d, %d)", this.row, this.col);
	}
	
	/**
	 * Recursive function for Creature to go North in Maze,
	 * edits maze and solPath on the way
	 */
	private boolean north() {
		
		boolean rightPath = false;
		
		if (this.maze.isClear(this.row - 1, this.col) ) {
			
			this.maze.markAsPath(this.row--, this.col);
			
			this.addToPath(DIRECTION.NORTH);
			
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.north();
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.east();
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.west();
			
			if (rightPath) {
				
				this.maze.markAsPath(this.row, this.col);
				
			} else {
				
				--this.solPathIndex;
				this.maze.markAsVisited(this.row++, this.col);
			}
		}
		
		return rightPath;
	}
	
	/**
	 * Recursive function for Creature to go South in Maze,
	 * edits maze and solPath on the way
	 */
	private boolean south() {
		
		boolean rightPath = false;
		
		if (this.maze.isClear(this.row + 1, this.col) ) {
			
			this.maze.markAsPath(this.row++, this.col);
			
			this.addToPath(DIRECTION.SOUTH);
			
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.south();
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.east();
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.west();
			
			if (rightPath) {
				
				this.maze.markAsPath(this.row, this.col);
				
			} else {

				--this.solPathIndex;
				this.maze.markAsVisited(this.row--, this.col);
			}
		}
		
		return rightPath;
	}
	
	/**
	 * Recursive function for Creature to go East in Maze,
	 * edits maze and solPath on the way
	 */
	private boolean east() {
		
		boolean rightPath = false;
		
		if (this.maze.isClear(this.row, this.col + 1) ) {
			
			this.maze.markAsPath(this.row, this.col++);
			
			this.addToPath(DIRECTION.EAST);
			
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.east();
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.north();
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.south();
			
			if (rightPath) {
				
				this.maze.markAsPath(this.row, this.col);
				
			} else {
			
				--this.solPathIndex;
				this.maze.markAsVisited(this.row, this.col--);
			}
		}
		
		return rightPath;
	}
	
	/**
	 * Recursive function for Creature to go West in Maze,
	 * edits maze and solPath on the way
	 */
	private boolean west() {
		
		boolean rightPath = false;
		
		if (this.maze.isClear(this.row, this.col - 1) ) {
			
			this.maze.markAsPath(this.row, this.col--);
			
			this.addToPath(DIRECTION.WEST);
			
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.west();
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.north();
			rightPath = (this.maze.isExit(this.row, this.col)) ? true : this.south();
			
			if (rightPath) {
				
				this.maze.markAsPath(this.row, this.col);
				
			} else {
				
				--this.solPathIndex;
				this.maze.markAsVisited(this.row, this.col++);
			}
		}
		
		return rightPath;
	}
	
	/**
	 * Uses most efficient method for building solPath
	 * @param dir The DIRECTION to add in the path
	 */
	private void addToPath(DIRECTION dir) {
		
		if (this.solPathIndex < this.solPath.length()) {
			
			this.solPath.setCharAt(this.solPathIndex++, dir.toChar());
			
		} else {
			
			++this.solPathIndex;
			this.solPath.append(dir.toChar());
		}
	}
	
	/**
	 * Returns path String in nice ass format
	 * @return formatted path String
	 */
	private String formatPath() {
		
		StringBuilder formatted = new StringBuilder();
		
		for (int i = 0; i < this.solPathIndex; ++i) {
			
			DIRECTION dir = DIRECTION.fromChar(this.solPath.charAt(i));
			
			int dirSteps = 1;
			
			while ((i < this.solPathIndex) && (dir.toChar() == this.solPath.charAt(i + 1))) {
				
				++i;
				++dirSteps;
			}
			
			formatted.append(dir.toChar());
			
			if (dirSteps > 1) {
				
				formatted.append('x');
				formatted.append(dirSteps);
			}
			
			if (i < this.solPathIndex - 1) {
				
				formatted.append('-');
			}
		}
		
		return formatted.toString();
	}
}
